### ISSUE TITLE
[BQ Cervantes]: help w/ input.

### ISSUE THREAD

**Issue by pazos** [koreader/koreader#4275]
* Device: BQ Cervantes 4.

I'm porting KOReader to BQ/Fnac devices. So far so good - repurposing most of kobo code - but I'm having problems w/ input. Device respond at touches (just fast touches, other longer touches are not recognized) but gestures and other stuff seems broken.

I've opened a forum thread on https://www.mobileread.com/forums/showthread.php?t=311571 that contains the output of evtest. A helping hand would be awesome.

BTW: I keep changes w/ current bq/fnac support on these branches:
for koreader -> https://github.com/pazos/koreader/tree/cervantes2
for koreader-base -> https://github.com/pazos/koreader-base/tree/cervantes2

**Comment by Frenzie** [comment]
That's very nice! (Sorry, can't be much help atm. @NiLuJe perhaps?)

**Comment by pazos** [comment]
It seems like a mix of singletouch and multitouch to me

```
Event: time 2469.393168, type 3 (EV_ABS), code 57 (ABS_MT_TRACKING_ID), value 144
Event: time 2469.393171, type 3 (EV_ABS), code 53 (ABS_MT_POSITION_X), value 296
Event: time 2469.393172, type 3 (EV_ABS), code 54 (ABS_MT_POSITION_Y), value 883
Event: time 2469.393175, type 3 (EV_ABS), code 52 (ABS_MT_ORIENTATION), value 0
Event: time 2469.393176, type 3 (EV_ABS), code 0 (ABS_X), value 883
Event: time 2469.393177, type 3 (EV_ABS), code 1 (ABS_Y), value 776
Event: time 2469.393179, type 1 (EV_KEY), code 330 (BTN_TOUCH), value 1
Event: time 2469.393180, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 1024
Event: time 2469.393181, -------------- SYN_REPORT ------------
Event: time 2469.455267, type 3 (EV_ABS), code 54 (ABS_MT_POSITION_Y), value 878
Event: time 2469.455268, type 3 (EV_ABS), code 48 (ABS_MT_TOUCH_MAJOR), value 6
Event: time 2469.455269, type 3 (EV_ABS), code 49 (ABS_MT_TOUCH_MINOR), value 6
Event: time 2469.455271, type 3 (EV_ABS), code 0 (ABS_X), value 878
Event: time 2469.455274, -------------- SYN_REPORT ------------
Event: time 2469.475476, type 3 (EV_ABS), code 57 (ABS_MT_TRACKING_ID), value -1
Event: time 2469.475479, type 1 (EV_KEY), code 330 (BTN_TOUCH), value 0
Event: time 2469.475480, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 0
Event: time 2469.475480, -------------- SYN_REPORT ------------
```

**single touch**
`ABS_X, ABS_Y  BTN_TOUCH`

**multitouch**
`ABS_MT_TRACKING_ID, ABS_MT_POSITION_X, ABS_MT_POSITION_Y, ABS_MT_ORIENTATION and ABS_PRESSURE`


BTN_TOUCH and ABS_PRESSURE go together (to value 1/1024 when the first finger touches the screen and to value 0/0 when the last finger leaves the screen). ABS_MT_TRACKING_ID happens at the same time too. This reports a growing number on touch and -1 after leave)

ABS_MT_ORIENTATION is always 0. ?¿

**Comment by NiLuJe** [comment]
Never really played with input despite my short foray with FBInk, so, not much help either, besides being able to show you what it looks like [on other devices](https://github.com/NiLuJe/FBInk/blob/4712758644ce3f841b32fb7ee86a832110d02443/fbink_button_scan.c#L201-L281) ;p.

**Comment by pazos** [comment]
@NiLuJe that helped.

My device acts like a phoenix and requires the ev_epoch_time tweak. That is the best possible combo. 

Touch events work but gestures are crazy. Every kind of swipe is detected as *west*. Some touches are detected as swipes too.

**Comment by pazos** [comment]
After a bit of hacking I'm going to implement a single touch protocol handler. We have one but doesn't handle devices that support both ST and MT(B). AFAICT BQ just ignore the mt_protocol completely and go ahead w/ ST.

The proposed solution (that works for me on my cervantes 4 and should work on Kobo ST devices) is:

```
-- Single Touch Protocol
function Input:handleTouchLegacyEv(ev)
    if ev.type == EV_ABS then
        if #self.MTSlots == 0 then
            table.insert(self.MTSlots, self:getMtSlot(self.cur_slot))
        end
        if ev.code == ABS_X then
            self:setCurrentMtSlot("x", ev.value)
        elseif ev.code == ABS_Y then
            self:setCurrentMtSlot("y", ev.value)
        elseif ev.code == ABS_PRESSURE then
            if ev.value ~= 0 then
                self:setCurrentMtSlot("id", 1)
            else
                self:setCurrentMtSlot("id", -1)
            end
        end
    elseif ev.type == EV_SYN then
        if ev.code == SYN_REPORT then
            for _, MTSlot in pairs(self.MTSlots) do
                self:setMtSlot(MTSlot.slot, "timev", TimeVal:new(ev.time))
            end

            -- feed ev in all slots to state machine
            local touch_ges = self.gesture_detector:feedEvent(self.MTSlots)
            self.MTSlots = {}
            if touch_ges then
                self:gestureAdjustHook(touch_ges)
                return Event:new("Gesture",
                    self.gesture_detector:adjustGesCoordinate(touch_ges)
                )
            end
        end
    end
end
```

This makes my device happy and swipes worked well. Opinions @NiLuJe & @Frenzie ?

**Comment by Frenzie** [comment]
Looks fine to me. (It's basically just a copy of the regular one, isn't it?)

**Comment by pazos** [comment]
@Frenzie: yes, with some cheating. On MT (w/ old Kobo ST support) we save ABS_X value in "abs_x" and ABS_Y value in "abs_y" and only store them on x/y when ABS_PRESSURE ~= 0 (using self:confirmAbsxy()). Here we write directly on x/y.

**Comment by frio** [comment]
This timing is serendipitous.  If you need someone to test on a Cervantes 3, please let me know.  I finally flashed the developer firmware the other day, and have since discovered the reader application crashes on page turn using the developer firmware :(.  After years employed in a "full stack" web shop, my C/systems programming fu is total rubbish, but I'm happy to test and collaborate where possible.

**Comment by Frenzie** [comment]
@frio I think it should be ready within a few days based on what @pazos said. You can track the progress on https://github.com/koreader/koreader-base/pull/745

**Comment by pazos** [comment]
@frio Good to know that somebody cares about running KOReader on Bq devices.

Yeah, what Frenzie said. Base support is almost there and should handle the Cervantes 3 too.  

And yeah, stock reader is totally broken so I didn't care about supporting connman to handle wifi and use wpa_supplicant instead. Switching between readers is possible and works well but after a few switches the stock application stalls and you need to reboot the device.

Please give us a couple of days. If you want to test it now please send me a mail and I will provide you a recent build which hopefully works on your device. (I would expect some wireless issues there).

**Comment by simoncarpinter** [comment]
Sure thing.  I've just switched back to the regular firmware so I can continue reading something I was close to finishing for now (... and probably shouldn't have been surprised by the fact that doing so is a factory reset, doh), but I'll be back on the developer firmware soon.  I'm happy to wait a few days :).

**Comment by frio** [comment]
Thanks @pazos.  I had some time today so I've spent it hacking around a bit.  After restoring the developer firmware and pivoting to an SD card boot so I can build a semi-useful on-device debugging environment, I've grabbed `koreader-cervantes-arm-cervantes-linux-gnueabi-v2018.11.1.zip` and tried to get things going.  I've run into a few issues; I'll raise a new bug in `koreader_base` and try and open a matching PR :), but I've hit a point where I've become stuck so may need some help.
